### PR TITLE
Updated Well Sampling

### PR DESCRIPTION
**Pull request by MattsonCam**
In this pr, I use the dp solution to select entire test wells for the non-minority groups until selecting additional test wells would change the minority group in the train/validation dataset. This consists of finding the largest number of cells that can be added for a given capacity of cells and number of wells, where cell capacities are restricted by well cell count. Please let me know if you have any questions or suggestions.

### CODE REVIEW

File: 1.train_models/utils/WellSubsetSum.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+                df_idx += 1
+
+                # Iterate through all possible capacities
+                # Each well must contain at least one cell
+                for w in np.arange(1, max_test_size + 1):
+
+                    # The largest number of wells at the current cell capacity (w) is
+                    # either at the previous largest number of cells,
+                    # or when adding the next wells. However, w
+                    # can not be less than the number of cells added from the well.
```
Comment (MikeLippincott) [inline 1.train_models/utils/WellSubsetSum.py]:
```suggestion
                    # cannot be less than the number of cells added from the well.
```

File: 1.train_models/utils/WellSubsetSum.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            test_wells.extend(groupdf.iloc[test_wells_int_idx][_well_col].tolist())
+
+            # Check that the number of test wells for the group isn't zero
+            if len(test_wells) == 0:
+                raise ValueError(f"The test well count of group {cat} is zero")
+
+            """
+            # Add wells to test set
+            test_wells.extend(groupdf[_well_col].tolist())
+            """
```
Comment (MikeLippincott) [inline 1.train_models/utils/WellSubsetSum.py]: Not sure if this code is still being used. If not, then consider removing

File: 1.train_models/utils/WellSubsetSum.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            The cell count column name.
+
+        _test_well_count: Integer
+            Number of wells to sample for testing.
+
+        Returns
+        -------
+        test_wells: List of Strings
+            Test well names.
+        """
```
Comment (MikeLippincott) [inline 1.train_models/utils/WellSubsetSum.py:39]: I really like this doc string. You do an excellent job describing the variables. Consider adding type hints.

**Review by MikeLippincott — APPROVED** [review]
This is a nice PR. LGTM

File: 1.train_models/utils/WellSubsetSum.py (hunk 1 above)
Comment (MattsonCam) [reply 1.train_models/utils/WellSubsetSum.py]: Changed

File: 1.train_models/utils/WellSubsetSum.py (hunk 2 above)
Comment (MattsonCam) [reply 1.train_models/utils/WellSubsetSum.py]: Not supposed to be there, thanks for catching that

**Comment by MattsonCam** [comment]
Thanks @MikeLippincott, merging now!

**Pull request merged**